pkg/api/publishWorker.go

```diff
@@ -0,0 +1,147 @@
+package api
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"github.com/xmtp/xmtpd/pkg/db"
+	"github.com/xmtp/xmtpd/pkg/db/queries"
+	"github.com/xmtp/xmtpd/pkg/registrant"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+)
+
+type PublishWorker struct {
+	ctx          context.Context
+	log          *zap.Logger
+	listener     <-chan []queries.StagedOriginatorEnvelope
+	notifier     chan<- bool
+	registrant   *registrant.Registrant
+	store        *sql.DB
+	subscription db.DBSubscription[queries.StagedOriginatorEnvelope]
+}
+
+func StartPublishWorker(
+	ctx context.Context,
+	log *zap.Logger,
+	reg *registrant.Registrant,
+	store *sql.DB,
+) (*PublishWorker, error) {
+	query := func(ctx context.Context, lastSeenID int64, numRows int32) ([]queries.StagedOriginatorEnvelope, int64, error) {
+		results, err := queries.New(store).SelectStagedOriginatorEnvelopes(
+			ctx,
+			queries.SelectStagedOriginatorEnvelopesParams{
+				LastSeenID: lastSeenID,
+				NumRows:    numRows,
+			},
+		)
+		if err != nil {
+			return nil, 0, err
+		}
+		if len(results) > 0 {
+			lastSeenID = results[len(results)-1].ID
+		}
+		return results, lastSeenID, nil
+	}
+	notifier := make(chan bool, 1)
+	subscription := db.NewDBSubscription(
+		ctx,
+		log,
+		query,
+		0, // lastSeenID
+		db.PollingOptions{Interval: time.Second, Notifier: notifier, NumRows: 100},
+	)
+	listener, err := subscription.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	worker := &PublishWorker{
+		ctx:          ctx,
+		log:          log,
+		notifier:     notifier,
+		subscription: *subscription,
+		listener:     listener,
+		registrant:   reg,
+		store:        store,
+	}
+	go worker.start()
+
+	return worker, nil
+}
+
+func (p *PublishWorker) NotifyStagedPublish() {
+	select {
+	case p.notifier <- true:
+	default:
+	}
+}
+
+func (p *PublishWorker) start() {
+	for {
+		select {
+		case <-p.ctx.Done():
+			close(p.notifier)
```

Comment:
Sending on a closed channel will cause a panic, so we want to make sure that the sender is in charge of closing the channel.

In this case, an API request might still come in after the context is canceled that calls `worker.NotifyStagedPublish`, which would cause a panic.

I'd suggest we make the notifier channel get passed in to the worker from the ApiService - and have the ApiService responsible for writing to the channel - so that the API service can be responsible for cleaning up in the right order during shutdown.

Reply:
Well, i guess it won't panic because `NotifyStagedPublish` sends in a select statement.

Still feels like we should try and close it from the sender side. That's almost always the right pattern in Go.

Reply:
If the channel is only getting closed when the server shuts down, maybe we don't need to close it at all.

Because you can have multiple concurrent senders it's going to be hard for the `ApiService` to know when it's fully safe to close. Given that only happens when the server is shutting down anyways, it should be fine.

Reply:
Good point, just not closing the channel at all. The nice thing about `NotifyStagedPublish` is that the notifier channel is self-contained in the `PublishWorker`, so an alternative would have been to check/close that channel inside that method instead